miner/scroll_worker.go

```diff
@@ -453,165 +507,196 @@ func (w *worker) startNewPipeline(timestamp int64) {
-				rawdb.WriteFirstQueueIndexNotInL2Block(w.eth.ChainDb(), startingHeader.ParentHash, overflowingL1MsgTx.QueueIndex+1)
+			if tx.IsL1MessageTx() {
+				txs.Shift()
 			} else {
-				w.prioritizedTx = nil
-				w.eth.TxPool().RemoveTx(res.OverflowingTx.Hash(), true)
-			}
-		} else if !res.OverflowingTx.IsL1MessageTx() {
-			// prioritize overflowing L2 message as the first txn next block
-			// no need to prioritize L1 messages, they are fetched in order
-			// and processed first in every block anyways
-			w.prioritizedTx = &prioritizedTransaction{
-				blockNumber: startingHeader.Number.Uint64() + 1,
-				tx:          res.OverflowingTx,
+				txs.Pop()
 			}
 		}
+	}
 
-		switch {
-		case res.OverflowingTx.IsL1MessageTx() &&
-			errors.Is(res.CCCErr, ccc.ErrBlockRowConsumptionOverflow):
-			l1TxRowConsumptionOverflowCounter.Inc(1)
-		case !res.OverflowingTx.IsL1MessageTx() &&
-			errors.Is(res.CCCErr, ccc.ErrBlockRowConsumptionOverflow):
-			l2TxRowConsumptionOverflowCounter.Inc(1)
-		case res.OverflowingTx.IsL1MessageTx() &&
-			errors.Is(res.CCCErr, ccc.ErrUnknown):
-			l1TxCccUnknownErrCounter.Inc(1)
-		case !res.OverflowingTx.IsL1MessageTx() &&
-			errors.Is(res.CCCErr, ccc.ErrUnknown):
-			l2TxCccUnknownErrCounter.Inc(1)
-		}
+	return false, nil
+}
+
+// processTxn
+func (w *worker) processTxn(tx *types.Transaction) (bool, error) {
+	if w.beforeTxHook != nil {
+		w.beforeTxHook()
 	}
 
-	var commitError error
-	if res.FinalBlock != nil {
-		if commitError = w.commit(res); commitError == nil {
-			return nil
-		}
-		log.Error("Commit failed", "header", res.FinalBlock.Header, "reason", commitError)
-		if _, isRetryable := commitError.(retryableCommitError); !isRetryable {
-			return commitError
-		}
+	// If we don't have enough gas for any further transactions then we're done
+	if w.current.gasPool.Gas() < params.TxGas {
+		return true, nil
 	}
-	w.startNewPipeline(time.Now().Unix())
-	return nil
+
+	// If we have collected enough transactions then we're done
+	// Originally we only limit l2txs count, but now strictly limit total txs number.
+	if !w.chain.Config().Scroll.IsValidTxCount(w.current.txs.Len() + 1) {
+		return true, nil
+	}
+
+	if tx.IsL1MessageTx() && tx.AsL1MessageTx().QueueIndex != w.current.nextL1MsgIndex {
+		// Continue, we might still be able to include some L2 messages
+		return false, ErrUnexpectedL1MessageIndex
+	}
+
+	if !tx.IsL1MessageTx() && !w.chain.Config().Scroll.IsValidBlockSize(w.current.blockSize+tx.Size()) {
+		// can't fit this txn in this block, silently ignore and continue looking for more txns
+		return false, errors.New("tx too big")
+	}
+
+	// Start executing the transaction
+	w.current.state.SetTxContext(tx.Hash(), w.current.txs.Len())
+
+	// create new snapshot for `core.ApplyTransaction`
+	snapState := w.current.state.Snapshot()
+	snapGasPool := *w.current.gasPool
+	snapGasUsed := w.current.header.GasUsed
+	snapCccLogger := w.current.cccLogger.Snapshot()
+
+	w.forceTestErr(tx)
```

Comment:
in production we should not invoke this?

Reply:
It should be fine, `forceTestErr` is a noop in production.